File: workflows/_lib/tests/gate_test.py
#!/usr/bin/env python3
"""Tests for workflows/_lib/gate.py (deterministic shell gate reports).

Run: python3 workflows/_lib/tests/gate_test.py

The verdict contract goes through the CLI rather than through run_gate, because the
exit code is half of that contract and only the process carries it.
"""

import importlib.util
import json
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

HERE = Path(__file__).resolve().parent
SCRIPT = HERE.parent / "gate.py"
_spec = importlib.util.spec_from_file_location("gate", SCRIPT)
assert _spec is not None and _spec.loader is not None
gate = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(gate)


def run_cli(*args: str) -> tuple[int, dict[str, object]]:

    with tempfile.TemporaryDirectory() as cwd:
        completed = subprocess.run(
            [sys.executable, str(SCRIPT), "--cwd", cwd, *args],
            capture_output=True,
            text=True,
            check=False,
        )
    return completed.returncode, json.loads(completed.stdout)


class TailTest(unittest.TestCase):
    def test_returns_whole_output_when_nothing_is_cut(self) -> None:
        self.assertEqual(gate.tail(b"alpha\nbeta\n", 100), "alpha\nbeta\n")

    def test_keeps_the_first_line_when_the_cut_lands_on_a_newline(self) -> None:
        self.assertEqual(gate.tail(b"alpha\nbeta\n", 5), "beta\n")

    def test_drops_a_first_line_the_cut_left_incomplete(self) -> None:
        self.assertEqual(gate.tail(b"alpha\nbeta\n", 8), "beta\n")

    def test_returns_empty_when_the_cut_leaves_no_complete_line(self) -> None:
        self.assertEqual(gate.tail(b"alphabeta", 4), "")

    def test_drops_a_partial_utf8_sequence_with_the_incomplete_line(self) -> None:
        # The cut lands inside the 3-byte U+3042, so the tail must start after the newline.
        data = "あいう\nおわり\n".encode()
        self.assertEqual(gate.tail(data, len(data) - 2), "おわり\n")

    def test_treats_a_carriage_return_as_a_line_end(self) -> None:
        self.assertEqual(gate.tail(b"alpha\rbeta", 8), "beta")


class ExactOutputLineTest(unittest.TestCase):
    def test_accepts_a_complete_line(self) -> None:
        self.assertTrue(gate.has_exact_output_line("not ok 1 - T-001 x\n", "", "not ok 1 - T-001 x"))

    def test_rejects_a_substring_of_a_line(self) -> None:
        self.assertFalse(gate.has_exact_output_line("not ok 1 - T-001 x\n", "", "T-001 x"))

    def test_matches_a_line_in_stderr(self) -> None:
        self.assertTrue(gate.has_exact_output_line("", "FAILED tests/a.py::t", "FAILED tests/a.py::t"))

    def test_rejects_evidence_carrying_a_newline(self) -> None:
        self.assertFalse(gate.has_exact_output_line("a\nb\n", "", "a\nb"))

    def test_rejects_empty_evidence(self) -> None:
        self.assertFalse(gate.has_exact_output_line("\n\n", "", ""))


class VerdictTest(unittest.TestCase):
    def test_passes_when_the_command_succeeds_as_expected(self) -> None:
        code, report = run_cli("--command", "printf 'done\\n'", "--expect", "pass")
        self.assertEqual(code, 0)
        self.assertEqual(report["verdict"], "pass")
        self.assertEqual(report["classification"], "pass")
        self.assertEqual(report["reason_codes"], [])
        self.assertIsNone(report["failure_route"])

    def test_fails_and_routes_when_a_pass_gate_sees_a_nonzero_exit(self) -> None:
        code, report = run_cli(
            "--command", "exit 3", "--expect", "pass", "--failure-route", "green:U-001"
        )
        self.assertEqual(code, 1)
        self.assertEqual(report["verdict"], "fail")
        self.assertEqual(report["reason_codes"], ["unexpected_failure"])
        self.assertEqual(report["failure_route"], "green:U-001")
        evidence = report["evidence"]
        assert isinstance(evidence, dict)
        self.assertEqual(evidence["exit_code"], 3)

    def test_blocks_a_red_gate_that_names_no_output_anchor(self) -> None:
        code, report = run_cli("--command", "exit 1", "--expect", "fail")
        self.assertEqual(code, 2)
        self.assertEqual(report["verdict"], "blocked")
        self.assertEqual(report["classification"], "usage_error")
        self.assertIn("--require-output", str(report["error"]))

    def test_confirms_a_red_gate_whose_anchor_is_a_complete_failure_line(self) -> None:
        code, report = run_cli(
            "--command", "printf 'ok 1 - T-001 x\\nnot ok 2 - T-002 y\\n'; exit 1",
            "--expect", "fail",
            "--require-output", "not ok 2 - T-002 y",
        )
        self.assertEqual(code, 0)
        self.assertEqual(report["verdict"], "pass")
        self.assertEqual(report["classification"], "expected_failure")

    def test_rejects_a_red_anchor_that_only_names_the_test(self) -> None:
        # "T-001 x" occurs inside the passing line, so an unrelated failure satisfies it.
        code, report = run_cli(
            "--command", "printf 'ok 1 - T-001 x\\nnot ok 2 - T-002 y\\n'; exit 1",
            "--expect", "fail",
            "--require-output", "T-001 x",
        )
        self.assertEqual(code, 1)
        self.assertEqual(report["verdict"], "fail")
        self.assertEqual(report["reason_codes"], ["missing_required_output"])

    def test_forbidden_output_matches_a_substring_rather_than_a_whole_line(self) -> None:
        code, report = run_cli(
            "--command", "printf 'warning: deprecated call\\n'",
            "--expect", "pass",
            "--forbid-output", "deprecated",
        )
        self.assertEqual(code, 1)
        self.assertEqual(report["reason_codes"], ["forbidden_output"])

    def test_blocks_and_reports_124_when_the_command_outruns_its_timeout(self) -> None:
        code, report = run_cli("--command", "sleep 5", "--expect", "pass", "--timeout-ms", "200")
        self.assertEqual(code, 124)
        self.assertEqual(report["verdict"], "blocked")
        self.assertEqual(report["reason_codes"], ["timeout"])
        evidence = report["evidence"]
        assert isinstance(evidence, dict)
        self.assertIs(evidence["timed_out"], True)


class UsageTest(unittest.TestCase):
    def test_rejects_an_unknown_flag(self) -> None:
        code, report = run_cli("--command", "true", "--expect", "pass", "--nope", "x")
        self.assertEqual(code, 2)
        self.assertIn("unknown argument: --nope", str(report["error"]))

    def test_rejects_a_repeated_singleton_flag(self) -> None:
        code, report = run_cli("--command", "true", "--command", "false", "--expect", "pass")
        self.assertEqual(code, 2)
        self.assertIn("only once", str(report["error"]))

    def test_rejects_a_relative_working_directory(self) -> None:
        completed = subprocess.run(
            [sys.executable, str(SCRIPT), "--cwd", "relative/dir", "--command", "true", "--expect", "pass"],
            capture_output=True,
            text=True,
            check=False,
        )
        self.assertEqual(completed.returncode, 2)
        self.assertIn("--cwd must be absolute", json.loads(completed.stdout)["error"])


class CalibrationTest(unittest.TestCase):
    def test_runs_a_red_command_without_an_anchor_and_marks_the_classification(self) -> None:
        code, report = run_cli(
            "--command", "printf 'not ok 1 - T-001 x\\n'; exit 1", "--calibrate"
        )
        self.assertEqual(code, 0)
        self.assertEqual(report["verdict"], "pass")
        self.assertEqual(report["classification"], "calibration_expected_failure")
        self.assertEqual(report["expected"], "fail")
        evidence = report["evidence"]
        assert isinstance(evidence, dict)
        self.assertEqual(evidence["stdout_tail"], "not ok 1 - T-001 x\n")

    def test_reports_a_calibration_whose_command_unexpectedly_passed(self) -> None:
        code, report = run_cli("--command", "printf 'all green\\n'", "--calibrate")
        self.assertEqual(code, 1)
        self.assertEqual(report["verdict"], "fail")
        self.assertEqual(report["classification"], "calibration_unexpected_pass")

    def test_refuses_an_anchor_because_calibration_is_what_discovers_one(self) -> None:
        code, report = run_cli("--command", "exit 1", "--calibrate", "--require-output", "x")
        self.assertEqual(code, 2)
        self.assertIn("takes no --require-output", str(report["error"]))

    def test_refuses_a_pass_expectation(self) -> None:
        code, report = run_cli("--command", "exit 1", "--calibrate", "--expect", "pass")
        self.assertEqual(code, 2)
        self.assertIn("--expect must be fail", str(report["error"]))

    def test_rejects_a_repeated_calibrate_flag(self) -> None:
        code, report = run_cli("--command", "exit 1", "--calibrate", "--calibrate")
        self.assertEqual(code, 2)
        self.assertIn("only once", str(report["error"]))


if __name__ == "__main__":
    unittest.main()

File: workflows/_lib/tests/gate.test.ts
/// <reference types="node" />
// Behavior tests for workflows/_lib/gate.ts, ported from workflows/_lib/tests/gate_test.py
// (the .py file stays until every scenario it covers has a home here; the ones already
// exercised as cross-checks live in workflows/_lib/tests/gate.differential.test.ts instead).
// gate.ts is a Red-step scaffold, so every test here is expected to fail against it until a
// later unit fills in the real parity implementation.
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const TS_SCRIPT = join(HERE, "..", "gate.ts");

interface GateRun {
  status: number | null;
  report: Record<string, unknown>;
}

function runCli(args: readonly string[]): GateRun {
  const result = spawnSync(process.execPath, [TS_SCRIPT, ...args], { encoding: "utf8" });
  return { status: result.status, report: JSON.parse(result.stdout) };
}

function withTempDir<T>(fn: (cwd: string) => T): T {
  const cwd = mkdtempSync(join(tmpdir(), "gate-test-"));
  try {
    return fn(cwd);
  } finally {
    rmSync(cwd, { recursive: true, force: true });
  }
}

test("T-010 未知フラグと重複 singleton と相対 cwd とアンカー無しの expect fail が blocked と usage_error になる", () => {
  withTempDir((cwd) => {
    const unknownFlag = runCli([
      "--cwd",
      cwd,
      "--command",
      "true",
      "--expect",
      "pass",
      "--nope",
      "x",
    ]);
    assert.equal(unknownFlag.status, 2, "unknown flag: exit code");
    assert.equal(unknownFlag.report.verdict, "blocked", "unknown flag: verdict");
    assert.equal(unknownFlag.report.classification, "usage_error", "unknown flag: classification");
    assert.match(String(unknownFlag.report.error), /unknown argument: --nope/);

    const duplicateSingleton = runCli([
      "--cwd",
      cwd,
      "--command",
      "true",
      "--command",
      "false",
      "--expect",
      "pass",
    ]);
    assert.equal(duplicateSingleton.status, 2, "duplicate singleton: exit code");
    assert.equal(duplicateSingleton.report.verdict, "blocked", "duplicate singleton: verdict");
    assert.equal(
      duplicateSingleton.report.classification,
      "usage_error",
      "duplicate singleton: classification",
    );
    assert.match(String(duplicateSingleton.report.error), /only once/);

    const noAnchor = runCli(["--cwd", cwd, "--command", "exit 1", "--expect", "fail"]);
    assert.equal(noAnchor.status, 2, "no anchor: exit code");
    assert.equal(noAnchor.report.verdict, "blocked", "no anchor: verdict");
    assert.equal(noAnchor.report.classification, "usage_error", "no anchor: classification");
    assert.match(String(noAnchor.report.error), /--require-output/);
  });

  const relativeCwd = runCli(["--cwd", "relative/dir", "--command", "true", "--expect", "pass"]);
  assert.equal(relativeCwd.status, 2, "relative cwd: exit code");
  assert.equal(relativeCwd.report.verdict, "blocked", "relative cwd: verdict");
  assert.equal(relativeCwd.report.classification, "usage_error", "relative cwd: classification");
  assert.match(String(relativeCwd.report.error), /--cwd must be absolute/);
});

test("T-011 切り詰めが残した先頭行が tail から落ち、完全な行が無ければ tail が空になる", () => {
  // gate.py's own test exercises tail() as a direct unit import. gate.ts's tsconfig
  // (workflows/../tsconfig.json, owned by U-001, out of this unit's file scope) sets no
  // allowImportingTsExtensions, so a same-repo .ts-to-.ts import of an internal helper fails
  // tsc's TS5097 while a Node-runtime-only extensionless specifier fails module resolution --
  // there is no specifier both accept. evidence.stdout_tail is tail()'s own output on the
  // CLI's surface, so the truncation behavior is observed through it instead.
  withTempDir((cwd) => {
    const incompleteFirstLineDropped = runCli([
      "--cwd",
      cwd,
      "--command",
      "printf 'alpha\\nbeta\\n'",
      "--expect",
      "pass",
      "--tail-bytes",
      "8",
    ]);
    const incompleteEvidence = incompleteFirstLineDropped.report.evidence as
      | Record<string, unknown>
      | undefined;
    assert.equal(
      incompleteEvidence?.stdout_tail,
      "beta\n",
      "the cut lands mid-line, so the incomplete first line is dropped",
    );

    const noCompleteLine = runCli([
      "--cwd",
      cwd,
      "--command",
      "printf 'alphabeta'",
      "--expect",
      "pass",
      "--tail-bytes",
      "4",
    ]);
    const noCompleteLineEvidence = noCompleteLine.report.evidence as
      | Record<string, unknown>
      | undefined;
    assert.equal(
      noCompleteLineEvidence?.stdout_tail,
      "",
      "no newline survives the cut, so there is no complete line to return",
    );
  });
});

test("T-012 完全な 1 行と一致しないアンカーが missing_required_output で落ち、禁止出力が forbidden_output で落ちる", () => {
  withTempDir((cwd) => {
    const nameOnlyAnchor = runCli([
      "--cwd",
      cwd,
      "--command",
      "printf 'ok 1 - T-001 x\\nnot ok 2 - T-002 y\\n'; exit 1",
      "--expect",
      "fail",
      "--require-output",
      "T-001 x",
    ]);
    assert.equal(nameOnlyAnchor.status, 1, "name-only anchor: exit code");
    assert.equal(nameOnlyAnchor.report.verdict, "fail", "name-only anchor: verdict");
    assert.deepEqual(nameOnlyAnchor.report.reason_codes, ["missing_required_output"]);

    const forbidden = runCli([
      "--cwd",
      cwd,
      "--command",
      "printf 'warning: deprecated call\\n'",
      "--expect",
      "pass",
      "--forbid-output",
      "deprecated",
    ]);
    assert.equal(forbidden.status, 1, "forbidden output: exit code");
    assert.equal(forbidden.report.verdict, "fail", "forbidden output: verdict");
    assert.deepEqual(forbidden.report.reason_codes, ["forbidden_output"]);
  });
});

test("T-013 calibrate がアンカー無しで走り classification に calibration_ が付き、アンカー指定と expect pass を拒否する", () => {
  withTempDir((cwd) => {
    const noAnchorRun = runCli([
      "--cwd",
      cwd,
      "--command",
      "printf 'not ok 1 - T-001 x\\n'; exit 1",
      "--calibrate",
    ]);
    assert.equal(noAnchorRun.status, 0, "calibrate without anchor: exit code");
    assert.equal(noAnchorRun.report.verdict, "pass", "calibrate without anchor: verdict");
    assert.equal(
      noAnchorRun.report.classification,
      "calibration_expected_failure",
      "calibrate without anchor: classification",
    );

    const anchoredCalibrate = runCli([
      "--cwd",
      cwd,
      "--command",
      "exit 1",
      "--calibrate",
      "--require-output",
      "x",
    ]);
    assert.equal(anchoredCalibrate.status, 2, "calibrate with anchor: exit code");
    assert.equal(anchoredCalibrate.report.verdict, "blocked", "calibrate with anchor: verdict");
    assert.match(String(anchoredCalibrate.report.error), /takes no --require-output/);

    const calibratePassExpectation = runCli([
      "--cwd",
      cwd,
      "--command",
      "exit 1",
      "--calibrate",
      "--expect",
      "pass",
    ]);
    assert.equal(calibratePassExpectation.status, 2, "calibrate with expect pass: exit code");
    assert.equal(
      calibratePassExpectation.report.verdict,
      "blocked",
      "calibrate with expect pass: verdict",
    );
    assert.match(String(calibratePassExpectation.report.error), /--expect must be fail/);
  });
});
